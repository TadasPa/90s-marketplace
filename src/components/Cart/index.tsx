import React, { FC, useCallback } from "react";
import ICartProduct from "../../types/cartProduct";

interface ICart {
  products: ICartProduct[];
  onRemove: (product: ICartProduct) => void;
}

const Cart: FC<ICart> = ({ products = [], onRemove }) => {
  const getItems = useCallback(
    () =>
      products.map((product) => {
        const handleRemove = () => onRemove(product);
        return (
          <li key={product.id}>
            <span>{product.title}</span>
            <span> Amount: {product.amount}</span>
            <button onClick={handleRemove}>x</button>
          </li>
        );
      }),
    [products, onRemove]
  );

  return <ul>{getItems()}</ul>;
};

export default Cart;
