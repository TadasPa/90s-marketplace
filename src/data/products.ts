import pictureA from '../assets/images/a.jpg';
import pictureB from '../assets/images/b.jpg';
import IProduct from '../types/product';

export const PRODUCT_A: IProduct = { id: 'a', title: 'Product A', price: '10USD', img: pictureA };
export const PRODUCT_B: IProduct = { id: 'b', title: 'Product B', price: '30USD', img: pictureB };
