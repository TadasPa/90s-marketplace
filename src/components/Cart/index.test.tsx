import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Cart from ".";
import { PRODUCT_A, PRODUCT_B } from "../../data/products";

test("renders Cart successfully with products", () => {
  const { container } = render(
    <Cart
      products={[
        { ...PRODUCT_A, amount: 5 },
        { ...PRODUCT_B, amount: 2 },
      ]}
      onRemove={jest.fn()}
    />
  );
  expect(container).toMatchSnapshot();
});

test("calls remove callback", () => {
  const handleRemoveMock = jest.fn();
  const removeButtonTestId = `remove-button-${PRODUCT_A.id}-testid`;
  const cartProduct = { ...PRODUCT_A, amount: 1 };

  render(<Cart products={[cartProduct]} onRemove={handleRemoveMock} />);

  userEvent.click(screen.getByTestId(removeButtonTestId));
  expect(handleRemoveMock).toHaveBeenCalledWith(cartProduct);
});
