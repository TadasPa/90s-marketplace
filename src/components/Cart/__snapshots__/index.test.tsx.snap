// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Cart successfully with products 1`] = `
<div>
  <ul>
    <li>
      <span>
        Product A
      </span>
      <span>
         Amount: 
        5
      </span>
      <button
        data-testid="remove-button-a-testid"
      >
        x
      </button>
    </li>
    <li>
      <span>
        Product B
      </span>
      <span>
         Amount: 
        2
      </span>
      <button
        data-testid="remove-button-b-testid"
      >
        x
      </button>
    </li>
  </ul>
</div>
`;
